spring-integration-r2dbc/src/main/java/org/springframework/integration/r2dbc/inbound/R2dbcMessageSource.java

```diff
@@ -0,0 +1,185 @@
+		Assert.notNull(queryExpression, "'queryExpression' must not be null");
+		this.databaseClient = databaseClient;
+		this.queryExpression = queryExpression;
+	}
+
+	/**
+	 * Provide a way to set the type of the entityClass that will be passed to the
+	 * {@link org.springframework.data.r2dbc.core.DatabaseClient#execute(String)}
+	 * method.
+	 * @param payloadType The t class.
+	 */
+	public void setPayloadType(Class<?> payloadType) {
+		Assert.notNull(payloadType, "'payloadType' must not be null");
+		this.payloadType = payloadType;
+	}
+
+	/**
+	 * Provide a way to manage which find* method to invoke on {@link R2dbcEntityOperations}.
+	 * Default is 'false', which means the {@link #receive()} method will use
+	 * the {@link org.springframework.data.r2dbc.core.DatabaseClient#execute(String)} method and will fetch all. If set
+	 * to 'true'{@link #receive()} will use {@link org.springframework.data.r2dbc.core.DatabaseClient#execute(String)}
+	 * and will fetch one and the payload of the returned {@link org.springframework.messaging.Message}
+	 * will be the returned target Object of type
+	 * identified by {@link #payloadType} instead of a List.
+	 * @param expectSingleResult true if a single result is expected.
+	 */
+	public void setExpectSingleResult(boolean expectSingleResult) {
+		this.expectSingleResult = expectSingleResult;
+	}
+
+	@Override
+	public String getComponentType() {
+		return "r2dbc:inbound-channel-adapter";
+	}
+
+	@Override
+	protected void onInit() {
+		this.evaluationContext = ExpressionUtils.createStandardEvaluationContext(getBeanFactory());
+		TypeLocator typeLocator = this.evaluationContext.getTypeLocator();
+		if (typeLocator instanceof StandardTypeLocator) {
+			/*
+			 * Register the R2dbc Query DSL package so they don't need a FQCN for QueryBuilder, for example.
+			 */
+			((StandardTypeLocator) typeLocator).registerImport("org.springframework.data.relational.core.query");
+		}
+		this.initialized = true;
+	}
+
+	/**
+	 * Execute a {@link Query} returning its results as the Message payload.
+	 * The payload can be either {@link reactor.core.publisher.Flux} or
+	 * {@link reactor.core.publisher.Mono} of objects of type identified by {@link #payloadType},
+	 * or a single element of type identified by {@link #payloadType}
+	 * based on the value of {@link #expectSingleResult} attribute which defaults to 'false' resulting
+	 * {@link org.springframework.messaging.Message} with payload of type
+	 * {@link reactor.core.publisher.Flux}. The collection name used in the
+	 */
+	@Override
+	protected Object doReceive() {
+		Assert.isTrue(this.initialized, "This class is not yet initialized. Invoke its afterPropertiesSet() method");
+		Mono<?> queryMono = Mono.fromSupplier(() -> this.queryExpression.getValue(this.evaluationContext));
+		if (this.expectSingleResult) {
+			return queryMono.flatMap(this::selectOne);
+		}
+		return queryMono.flatMapMany(this::select);
+	}
+
+	private Mono<?> selectOne(Object queryMono) {
```

Comment:
Well, its not `Mono` here any more. Isn't it? 
I mean the name for the param is misleading 😄